### PR TITLE
fix(assertions): throw error or warn when `synth` is called multiple times on mutated construct tree

### LINKED ISSUE #24689: aws-cdk-lib/assertions: `Unable to find artifact with id` when creating two instances of the same Stack in Jest test
### Describe the bug

I'm migrating from CDK v1 to CDK v2. Have fixed all the syntax issues and now fixing unit tests.

In V1 tests I have the following code which I migrated to v2:

```
describe('Website Stack', () => {
  const app = new App();
  const envOregon = { region: 'us-west-2' };

  const stages = [
    {
      stage: STAGE.PROD,
    },
    {
      stage: STAGE.BETA,
    },
  ];

  stages.forEach(({ stage }) => {
    const stack = new WebsiteStack(app, `${stage}WebsiteStack`, {
      env: envOregon,
      subdomain: stage,
    });

    // Here it fails with `Unable to find artifact with id`
    const template = Template.fromStack(stack);
```

It fails to find the the second instance on the last line with `Template.fromStack()`. It always can't find the second stack even when I change the order of the stack definition.

### Expected Behavior

Previously in V1 the test used to work by finding the second stack as well.

### Current Behavior

Internally I see that `Template.fromStack()` calls the following code:

    return new Template(toTemplate(stack), templateParsingOptions);

... which in turn calls:

```
return stack.nestedStackParent
    ? JSON.parse(fs.readFileSync(path.join(assembly.directory, stack.templateFile)).toString('utf-8'))
    : assembly.getStackArtifact(stack.artifactId).template;
```

In the above code `fs.readFileSync(path.join(assembly.directory, stack.templateFile))` fails because for some reason the test does not compile/synthesize the second instance. In the `assembly.directory` I only see the first instance, whichever is the first.

```
$ ls -lat "/private/var/folders/l7/p9ksylw17x34lc5_xstwndmr0000gp/T/cdk.out4nhobd"
cdk.out
manifest.json
tree.json
WebsiteStack.assets.json
WebsiteStack.template.json
```

The directory has no template for the `BetaWebsiteStack`.

### Reproduction Steps

- Followed the migration guide to move to CDK v2.
- All syntax errors are fixed.
- All deprecated constructs are migrated.
- Run `npm test` in CDK v2 project.

### Possible Solution

I managed to fix the test by creating a separate instance of the `App` for each stack as follows:

```
describe('Website Stack', () => {
  const envOregon = { region: 'us-west-2' };

  const stages = [
    {
      stage: STAGE.PROD,
    },
    {
      stage: STAGE.BETA,
    },
  ];

  stages.forEach(({ stage }) => {
    const app = new App();

    const stack = new WebsiteStack(app, `${stage}WebsiteStack`, {
      env: envOregon,
      subdomain: stage,
    });

    // Here it fails with `Unable to find artifact with id`
    const template = Template.fromStack(stack);
```

### Additional Information/Context

_No response_

### CDK CLI Version

2.69.0 (build 60a5b2a)

### Framework Version

2.69.0

### Node.js Version

v16.15.0

### OS

macOS 13.2.1

### Language

Typescript

### Language Version

5.0.2

### Other information

_No response_

### PR DESCRIPTION
**Pull request by sumupitchayan**
Closes #24689 

### Reason for this change

Calling `Template.fromStack(stack)` twice on the same stack object will throw the error `Unable to find artifact with id` if the stack is mutated after the first `Template.fromStack(stack)` call. This is because synth should only be called once - from the comment on the `synth` function:

> _Once an assembly has been synthesized, it cannot be modified. Subsequent calls will return the same assembly._

Second call of `Template.fromStack(stack)` tries to find the mutated stack since `app.synth()` caches and returns the assembly from the first synth call.

### Description of changes

This PR checks to see whether or not the construct tree has been each time `synth()` is called - if it has, then we either throw an error or a warning. We will only throw a warning if synth is being called from `process.once('beforeExit')`.

### Description of how you validated changes

Unit test with the customer's same example case, asserting than an error is thrown when synth is called twice after a stack has been mutated after the first synth call.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/pipelines/test/compliance/synths.test.ts [hunk 1]
```diff
@@ -186,6 +186,27 @@ test('CodeBuild: environment variables specified in multiple places are correctl
     }),
   });
 
+  new ModernTestGitHubNpmPipeline(pipelineStack, 'Cdk-2', {
```
Comment (sumupitchayan) [inline packages/aws-cdk-lib/pipelines/test/compliance/synths.test.ts:189]: This existing pipelines test had to be changed to not throw an error - all I did was move the stack mutations before the `Template.fromStack()` calls

File: packages/aws-cdk-lib/core/lib/stage.ts [hunk 2]
```diff
@@ -210,13 +217,52 @@ export class Stage extends Construct {
...
+        }
+      }
+    }
+
+    // If the construct paths set has changed
+    if (!constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      if (options.warnInsteadOfError) {
+        Annotations.of(this).addWarningV2('@aws-cdk/core:synth-warning', 'Construct paths have been modified since the last synth call.');
+      } else {
+        throw new Error('Synth was called since the last call.');
```
Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:249]: This error doesn't really say anything. Error message should be clear and concise with and actionable item to fix the issue.

File: packages/aws-cdk-lib/core/lib/stage.ts [hunk 3]
```diff
@@ -210,13 +217,52 @@ export class Stage extends Construct {
...
+        if (!Stage.isStage(child)) {
+          recurseAndlistAllConstructPaths(child);
+        }
+      }
+    }
+
+    // If the construct paths set has changed
+    if (!constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      if (options.warnInsteadOfError) {
+        Annotations.of(this).addWarningV2('@aws-cdk/core:synth-warning', 'Construct paths have been modified since the last synth call.');
```
Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:247]: Similar to the error, I don't really know what this warning is telling me and why it is a warning.

File: packages/aws-cdk-lib/core/lib/stage.ts (hunk 2 above)
Comment (sumupitchayan) [reply packages/aws-cdk-lib/core/lib/stage.ts:249]: Yup, rewriting these warning/error messages are my last TODO's here :) - will take it out of draft after

File: packages/aws-cdk-lib/pipelines/test/compliance/synths.test.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/aws-cdk-lib/pipelines/test/compliance/synths.test.ts:189]: That's perfect!

File: packages/aws-cdk-lib/core/lib/stage.ts [hunk 4]
```diff
@@ -210,16 +216,62 @@ export class Stage extends Construct {
...
         skipValidation: options.skipValidation,
         validateOnSynthesis: options.validateOnSynthesis,
       });
+      newConstructPaths = this.listAllConstructPaths(this);
+      this.constructPathsCache = newConstructPaths;
     }
 
+    // If the construct paths set has changed
+    if (!this.constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      const errorMessage = 'The construct tree has been modified after synthesis. Only the results of the first synth() call are written to disk, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called.';
```
Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:234]: "written to disk" still feels to implentation-detaily for me.

```suggestion
      const errorMessage = 'The construct tree has been modified after synthesis. Only the results of the first synth() call are used, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called.';
```

File: packages/aws-cdk-lib/core/lib/stage.ts [hunk 5]
```diff
@@ -210,16 +216,62 @@ export class Stage extends Construct {
...
+      if (options.errorOnDuplicateSynth ?? true) {
+        throw new Error(errorMessage);
+      } else {
+        // eslint-disable-next-line no-console
+        console.error(errorMessage);
```
Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:239]: So we print this as a warning. And now what? The message text implies that we are in the warning situation and is trying to say the following:

- Hey we have detected that you called synth() multiple and changed the construct tree after the first synth()
- We are going to ignoring all changes, but are not failing the execution.
- You should stop doing this, unless you really know what you do.

File: packages/aws-cdk-lib/core/lib/stage.ts [hunk 6]
```diff
@@ -210,16 +216,62 @@ export class Stage extends Construct {
...
       });
+      newConstructPaths = this.listAllConstructPaths(this);
+      this.constructPathsCache = newConstructPaths;
     }
 
+    // If the construct paths set has changed
+    if (!this.constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      const errorMessage = 'The construct tree has been modified after synthesis. Only the results of the first synth() call are written to disk, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called.';
+      if (options.errorOnDuplicateSynth ?? true) {
+        throw new Error(errorMessage);
```
Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:236]: Not sure we can really use the exact same message. the call to action is different when there is an error vs a warning. In my head, the error scenario is this:

- Hey we have detected that you called synth() multiple times and changed the construct tree after the first synth()
- You cannot do this, we are failing the execution.
- If you really know what you are doing, you can set this flag to keep going: <enter flag here>

**Review by rix0rrr — APPROVED** [review]

File: packages/aws-cdk-lib/core/lib/stage.ts [hunk 7]
```diff
@@ -210,16 +216,61 @@ export class Stage extends Construct {
...
+      this.constructPathsCache = newConstructPaths;
     }
 
+    // If the construct paths set has changed
+    if (!this.constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      if ((options.errorOnDuplicateSynth ?? true) && !options.force) {
+        throw new Error('Synthesis has been called multiple times and the construct tree was modified after the first synthesis. This is not allowed: if you intended to do this, then call synthesis with the --force flag set to true.');
+      } else {
+        // eslint-disable-next-line no-console
+        console.error('Synthesis has been called mulitple times and the construct tree was modified after the first synthesis. Only the results of the first synth() call are used, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called unless this is intentional.');
```
Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:238]: Why are we not using a construct warning here?

Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:238]:
```suggestion
        console.error('Synthesis has been called multiple times and the construct tree was modified after the first synthesis. Only the results of the first synth() call are used, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called unless this is intentional.');
```

Comment (sumupitchayan) [reply packages/aws-cdk-lib/core/lib/stage.ts]: Rico's suggestion: https://github.com/aws/aws-cdk/pull/31865#discussion_r1819425777

File: packages/aws-cdk-lib/core/lib/stage.ts [hunk 8]
```diff
@@ -210,16 +216,62 @@ export class Stage extends Construct {
...
     }
 
+    // If the construct paths set has changed
+    if (!this.constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      const errorMessage = 'Synthesis has been called multiple times and the construct tree was modified after the first synthesis.';
+      if ((options.errorOnDuplicateSynth ?? true) && !options.force) {
+        throw new Error(errorMessage + ' This is not allowed.');
+      } else {
+        // eslint-disable-next-line no-console
+        console.error(errorMessage + ' Only the results of the first synth() call are used, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called unless this is intentional.');
```
Comment (mrgrain) [inline packages/aws-cdk-lib/core/lib/stage.ts:239]: Why are we not using the regular warning mechanism?

Comment (sumupitchayan) [reply packages/aws-cdk-lib/core/lib/stage.ts:239]: Rico's suggestion: https://github.com/aws/aws-cdk/pull/31865#discussion_r1819425777

**Review by mrgrain — APPROVED** [review]
Conditionally approved until you've checked that `addWarningV2()` really doesn't work.

**Comment by sumupitchayan** [comment]
> Conditionally approved until you've checked that `addWarningV2()` really doesn't work.

Just tested and confirmed that `addWarningV2()` doesn't work - going with Rico's `console.error()` suggestion

**Pull request merged**